# Comparison of functional normalization implementations

## Load example data set 

Retrieve the data from the Gene Expression Omnibus (GEO) website
(http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE55491).

```{r}
dir.create(path <- "data")
if (length(list.files(path, "*.idat$")) == 0) {
  filename <-  file.path(path, "gse55491.tar")
  download.file("http://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE55491&format=file", filename)
  cat(date(), "Extracting files from GEO archive.\n")
  system(paste("cd", path, ";", "tar xvf", basename(filename)))
  unlink(filename)
  cat(date(), "Unzipping IDAT files.\n")
  system(paste("cd", path, ";", "gunzip *.idat.gz"))
}
```

## Load and normalize using `meffil`
Load the code, probe annotation and sample filename information.
```{r}
library(meffil)
probes <- meffil.probe.info()
basenames <- meffil.basenames(path)
```

Next we collect controls, perform background and dye bias correction and then
compute quantiles for each sample
using two different dye intensity targets, one very low and the other very high
to see if the resulting beta values are affected.
```{r}
norm.objects.lo <- mclapply(basenames, function(basename) {
    meffil.compute.normalization.object(basename, probes=probes, dye.intensity=500) 
})
norm.objects.hi <- mclapply(basenames, function(basename) {
    meffil.compute.normalization.object(basename, probes=probes, dye.intensity=5000) 
})
```

Normalize quantiles from each sample using the control
matrix to identify batch effects.
```{r}
norm.objects.lo <- meffil.normalize.objects(norm.objects.lo, number.pcs=2)
norm.objects.hi <- meffil.normalize.objects(norm.objects.hi, number.pcs=2)
```

Apply quantile normalization to methylation levels of each sample.
```{r}
B.lo <- do.call(cbind, mclapply(norm.objects.lo, function(object) {
    meffil.get.beta(meffil.normalize.sample(object, probes)) 
}))
B.hi <- do.call(cbind, mclapply(norm.objects.hi, function(object) {
    meffil.get.beta(meffil.normalize.sample(object, probes)) 
})) 
```

```{r}
quantile(B.lo - B.hi)
```
