From 9034dd092acca0fd93e8c82a22064a16e202554a Mon Sep 17 00:00:00 2001
From: Matthew Suderman <matthewsuderman@hotmail.com>
Date: Sat, 21 Feb 2015 15:03:48 +0000
Subject: [PATCH] fixes control matrix error

---
 R/preprocessFunnorm.R |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/R/preprocessFunnorm.R b/R/preprocessFunnorm.R
index c7f3ae0..9dcba4d 100644
--- a/R/preprocessFunnorm.R
+++ b/R/preprocessFunnorm.R
@@ -196,7 +196,7 @@ preprocessFunnorm <- function(rgSet, nPCs=2, sex = NULL, bgCorr = TRUE, dyeCorr
     ## Bisulfite conversion extraction for probe type I:
     index <- match("BISULFITE CONVERSION I", controlNames)
     addr <- getCtrlsAddr(exType = sprintf("BS Conversion I%sC%s", c(" ", "-", "-"), 1:3), index = index)
-    greenControls.current <- redControls[[ index ]][addr,]
+    greenControls.current <- greenControls[[ index ]][addr,]
     addr <- getCtrlsAddr(exType = sprintf("BS Conversion I-C%s", 4:6), index = index)
     redControls.current <- redControls[[ index ]][addr,]
     bisulfite1 <- colMeans(redControls.current + greenControls.current, na.rm = TRUE)
-- 
1.7.3

