<!DOCTYPE html>
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>

<title>Normalize an EPIC v2 dataset</title>

<link rel="stylesheet" href="https://cdn.jsdelivr.net/gh/highlightjs/cdn-release@11.6.0/build/styles/github.min.css">
<script src="https://cdn.jsdelivr.net/combine/gh/highlightjs/cdn-release@11.6.0/build/highlight.min.js,npm/@xiee/utils/js/load-highlight.js" async></script>



<style type="text/css">
body, td {
  font-family: sans-serif;
  background-color: white;
  font-size: 13px;
}
body {
  max-width: 800px;
  margin: auto;
  padding: 1em;
  line-height: 1.5;
}
tt, code, pre {
  font-family: 'DejaVu Sans Mono', 'Droid Sans Mono', 'Lucida Console', Consolas, Monaco, monospace;
}
a:visited { color: #80007f; }
pre, img { max-width: 100%; }
code {
  font-size: 92%;
  border: 1px solid #ccc;
}
code[class] { background-color: #F8F8F8; }
code.language-undefined { background-color: inherit; }
table {
  margin: auto;
  border-top: 1px solid #666;
  border-bottom: 1px solid #666;
}
table thead th { border-bottom: 1px solid #ddd; }
th, td { padding: 5px; }
thead, tfoot, tr:nth-child(even) { background: #eee; }
blockquote {
  color:#666;
  margin:0;
  padding-left: 1em;
  border-left: 0.5em #eee solid;
}
hr { border: 1px #ddd dashed; }

@media print {
  * {
    background: transparent !important;
    color: black !important;
    filter:none !important;
  }
  body {
    font-size: 12pt;
    max-width: 100%;
  }
  a, a:visited { text-decoration: underline; }
  hr {
    visibility: hidden;
    page-break-before: always;
  }
  pre, blockquote {
    padding-right: 1em;
    page-break-inside: avoid;
  }
  tr, img { page-break-inside: avoid; }
  img { max-width: 100% !important; }
  @page :left { margin: 15mm 20mm 15mm 10mm; }
  @page :right { margin: 15mm 10mm 15mm 20mm; }
  p, h2, h3 { orphans: 3; widows: 3; }
  h2, h3 { page-break-after: avoid; }
}
</style>



</head>

<body>
<h1>Normalize an EPIC v2 dataset</h1>
<h2>Download example data set</h2>
<p>Information about the new EPIC v2 microarray can be obtained here:</p>
<p><a href="https://www.illumina.com/products/by-type/microarray-kits/infinium-methylation-epic.html">https://www.illumina.com/products/by-type/microarray-kits/infinium-methylation-epic.html</a></p>
<pre><code class="language-r">download.epic2.demo.dataset &lt;- function() {
    dir.create(path &lt;- &quot;data-epic2-demo&quot;)
    
    if (length(list.files(path, &quot;*.idat$&quot;)) == 0) {
        download.zip &lt;- function(url, path) {
            filename &lt;- file.path(path, &quot;data.zip&quot;)
            download.file(url, filename)
            filenames &lt;- unzip(filename, junkpaths=T, exdir=path)
            unlink(filename)
            invisible(filenames)
        }    

        url &lt;- &quot;https://support.illumina.com/content/dam/illumina-support/documents/downloads/productfiles/methylationepic/DemoDataEPIC_v2.zip&quot;
        download.zip(url, file.path(path, &quot;data.zip&quot;))
    }
    
    path
}
</code></pre>
<pre><code class="language-r">path &lt;- download.epic2.demo.dataset()
</code></pre>
<h2>Normalize dataset</h2>
<p>Create samplesheet</p>
<pre><code class="language-r">library(meffil)
samplesheet &lt;- meffil.read.samplesheet(base=path, pattern=&quot;SampleSheet&quot;)
</code></pre>
<pre><code>## [read.450k.sheet] Found the following CSV files:
## [1] &quot;data-epic2-demo/Demo_EPIC-8v2-0_A1_SampleSheet_16.csv&quot;
</code></pre>
<p>Parameters.</p>
<pre><code class="language-r">qc.file &lt;- &quot;epic2-demo/qc-report.html&quot;
author &lt;- &quot;Illumina, et al.&quot;
study &lt;- &quot;EPIC demo dataset&quot;
number.pcs &lt;- 2
norm.file &lt;- &quot;epic2-demo/normalization-report.html&quot;
</code></pre>
<p>Generate QC objects.</p>
<pre><code class="language-r">qc.objects &lt;- meffil.qc(samplesheet, cell.type.reference=&quot;blood gse35069 complete&quot;, verbose=T)
</code></pre>
<pre><code>## [read.idat] Mon Apr 17 00:40:23 2023 Reading data-epic2-demo/206891110001_R01C01_Grn.idat 
## [read.idat] Mon Apr 17 00:40:23 2023 Reading data-epic2-demo/206891110001_R01C01_Red.idat 
## [extract.detection.pvalues] Mon Apr 17 00:40:27 2023  
## [extract.beadnum] Mon Apr 17 00:40:31 2023  
## [extract.snp.betas] Mon Apr 17 00:40:35 2023  
## [extract.controls] Mon Apr 17 00:40:35 2023  
## [background.correct] Mon Apr 17 00:40:37 2023 background correction for dye = R 
## [background.correct] Mon Apr 17 00:40:39 2023 background correction for dye = G 
## [dye.bias.correct] Mon Apr 17 00:40:41 2023  
## [FUN] Mon Apr 17 00:40:46 2023 predicting sex
</code></pre>
<pre><code class="language-r">## If you get an error relating to a function from
## the 'preprocessCore' R package, 
## then you may need to reinstall it as follows: 
## BiocManager::install(&quot;preprocessCore&quot;, configure.args=&quot;--disable-threading&quot;, force = TRUE)
</code></pre>
<p>QC report.</p>
<pre><code class="language-r">qc.summary &lt;- meffil.qc.summary(qc.objects, verbose=T)
</code></pre>
<pre><code>## [meffil.qc.summary] Mon Apr 17 00:49:13 2023 Sex summary TRUE 
## [meffil.qc.summary] Mon Apr 17 00:49:13 2023 Meth vs unmeth summary 
## [meffil.qc.summary] Mon Apr 17 00:49:13 2023 Control means summary 
## [meffil.qc.summary] Mon Apr 17 00:49:13 2023 Sample detection summary 
## [meffil.qc.summary] Mon Apr 17 00:49:15 2023 CpG detection summary 
## [meffil.qc.summary] Mon Apr 17 00:49:15 2023 Sample bead numbers summary 
## [meffil.qc.summary] Mon Apr 17 00:49:16 2023 CpG bead numbers summary 
## [meffil.qc.summary] Mon Apr 17 00:49:17 2023 Cell count summary 
## [meffil.qc.summary] Mon Apr 17 00:49:17 2023 Genotype concordance
</code></pre>
<pre><code class="language-r">meffil.qc.report(
    qc.summary,
    output.file=qc.file,
    author=author,
    study=study)
</code></pre>
<pre><code>## [meffil.qc.report] Mon Apr 17 00:49:18 2023 Writing report as html file to epic2-demo/qc-report.html
</code></pre>
<p>Normalization dataset.</p>
<pre><code class="language-r">norm.objects &lt;- meffil.normalize.quantiles(
    qc.objects,
    number.pcs=number.pcs,
    verbose=T)
</code></pre>
<pre><code>## [meffil.normalize.quantiles] Mon Apr 17 00:41:33 2023 selecting dye correction reference 
## [meffil.normalize.quantiles] Mon Apr 17 00:41:33 2023 creating control matrix 
## [meffil.normalize.quantiles] Mon Apr 17 00:41:33 2023 normalizing quantiles 
## [FUN] Mon Apr 17 00:41:33 2023 genomic.iG M 
## [FUN] Mon Apr 17 00:41:33 2023 genomic.iG U 
## [FUN] Mon Apr 17 00:41:33 2023 genomic.iR M 
## [FUN] Mon Apr 17 00:41:33 2023 genomic.iR U 
## [FUN] Mon Apr 17 00:41:33 2023 genomic.ii M 
## [FUN] Mon Apr 17 00:41:33 2023 genomic.ii U 
## [FUN] Mon Apr 17 00:41:33 2023 autosomal.iG M 
## [FUN] Mon Apr 17 00:41:33 2023 autosomal.iG U 
## [FUN] Mon Apr 17 00:41:33 2023 autosomal.iR M 
## [FUN] Mon Apr 17 00:41:33 2023 autosomal.iR U 
## [FUN] Mon Apr 17 00:41:33 2023 autosomal.ii M 
## [FUN] Mon Apr 17 00:41:33 2023 autosomal.ii U 
## [FUN] Mon Apr 17 00:41:33 2023 not.y.iG M 
## [FUN] Mon Apr 17 00:41:33 2023 not.y.iG U 
## [FUN] Mon Apr 17 00:41:33 2023 not.y.iR M 
## [FUN] Mon Apr 17 00:41:33 2023 not.y.iR U 
## [FUN] Mon Apr 17 00:41:33 2023 not.y.ii M 
## [FUN] Mon Apr 17 00:41:33 2023 not.y.ii U 
## [FUN] Mon Apr 17 00:41:34 2023 sex M 
## [FUN] Mon Apr 17 00:41:34 2023 sex U 
## [FUN] Mon Apr 17 00:41:34 2023 chrx M 
## [FUN] Mon Apr 17 00:41:34 2023 chrx U 
## [FUN] Mon Apr 17 00:41:34 2023 chry M 
## [FUN] Mon Apr 17 00:41:34 2023 chry U
</code></pre>
<pre><code class="language-r">beta.meffil &lt;- meffil.normalize.samples(
    norm.objects,
    just.beta=T, 
    cpglist.remove=qc.summary$bad.cpgs$name,
    verbose=T)
</code></pre>
<pre><code>## [read.idat] Mon Apr 17 00:41:34 2023 Reading data-epic2-demo/206891110001_R01C01_Grn.idat 
## [read.idat] Mon Apr 17 00:41:34 2023 Reading data-epic2-demo/206891110001_R01C01_Red.idat 
## [background.correct] Mon Apr 17 00:41:34 2023 background correction for dye = R 
## [background.correct] Mon Apr 17 00:41:37 2023 background correction for dye = G 
## [dye.bias.correct] Mon Apr 17 00:41:39 2023  
## [meffil.normalize.sample] Mon Apr 17 00:41:42 2023 Normalizing methylated and unmethylated signals.
</code></pre>
<p>Compute principal components of the normalized methylation matrix.</p>
<pre><code class="language-r">pcs &lt;- meffil.methylation.pcs(
    beta.meffil,
    sites=meffil.get.autosomal.sites(&quot;epic&quot;),
    verbose=T)
</code></pre>
<pre><code>## [meffil.methylation.pcs] Mon Apr 17 00:42:11 2023 Calculating CpG variance 
## [meffil.methylation.pcs] Mon Apr 17 00:42:12 2023 Calculating beta PCs
</code></pre>
<p>Normalization report.</p>
<pre><code class="language-r">parameters &lt;- meffil.normalization.parameters(norm.objects)
parameters$batch.threshold &lt;- 0.01
norm.summary &lt;- meffil.normalization.summary(
    norm.objects=norm.objects,
    pcs=pcs,
    parameters=parameters,
    verbose=T)
</code></pre>
<pre><code>## [meffil.plot.control.batch] Mon Apr 17 00:49:28 2023 Extracting batch variables 
## [meffil.plot.control.batch] Mon Apr 17 00:49:28 2023 Testing associations 
## [meffil.plot.probe.batch] Mon Apr 17 00:49:33 2023 Extracting batch variables 
## [meffil.plot.probe.batch] Mon Apr 17 00:49:33 2023 Testing associations
</code></pre>
<pre><code class="language-r">meffil.normalization.report(
    norm.summary,
    output.file=norm.file,
    author=author,
    study=study)
</code></pre>
<pre><code>## [meffil.normalization.report] Mon Apr 17 00:49:39 2023 Writing report as html file to epic2-demo/normalization-report.html
</code></pre>
<h2>Horvath’s clock and the EPIC v2 microarray</h2>
<pre><code class="language-r">#require(RCurl)
#clock &lt;- read.csv(textConnection(getURL(&quot;https://labs.genetics.ucla.edu/horvath/dnamage/AdditionalFile3.csv&quot;)), comment.char=&quot;#&quot;, stringsAsFactors=F)
clock &lt;- read.csv(&quot;AdditionalFile3.csv&quot;, comment.char=&quot;#&quot;, stringsAsFactors=F)
length(setdiff(clock$CpGmarker, rownames(beta.meffil)))
</code></pre>
<pre><code>## [1] 14
</code></pre>
<pre><code class="language-r">nrow(clock)
</code></pre>
<pre><code>## [1] 354
</code></pre>
<pre><code class="language-r">length(intersect(clock$CpGmarker, rownames(beta.meffil)))/nrow(clock)
</code></pre>
<pre><code>## [1] 0.960452
</code></pre>
<p>Some of the ‘clock’ sites are missing.</p>
<h2>Hannum predictor CpG sites and the EPIC microarray</h2>
<p>71 CpG sites were used in the Hannum et al. age predictor:</p>
<blockquote>
<p>Hannum G, et al.
Genome-wide methylation profiles reveal quantitative views of human aging rates.
Mol Cell. 2013 Jan 24;49(2):359-67.</p>
</blockquote>
<p>The list can be obtained from here:
<a href="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC3780611/bin/NIHMS418935-supplement-02.xlsx">http://www.ncbi.nlm.nih.gov/pmc/articles/PMC3780611/bin/NIHMS418935-supplement-02.xlsx</a></p>
<pre><code class="language-r">hannum.sites &lt;- readLines(&quot;hannum.txt&quot;)
length(hannum.sites)
</code></pre>
<pre><code>## [1] 71
</code></pre>
<pre><code class="language-r">length(setdiff(hannum.sites, rownames(beta.meffil)))
</code></pre>
<pre><code>## [1] 7
</code></pre>
<pre><code class="language-r">length(intersect(hannum.sites, rownames(beta.meffil)))/length(hannum.sites)
</code></pre>
<pre><code>## [1] 0.9014085
</code></pre>
<p>Some but not all of the sites are missing.</p>
<h2>Duplicated probes</h2>
<p>Several thousand probes appear in duplicate across the microarray.
By default, the <code>meffil.normalize.samples</code> function collapses
duplicated probes by taking their median.
It is possible to a different summarizing function by setting the
<code>dup.fun</code> argument.
It is also possible to keep the duplicate probes by
setting <code>dup.fun</code> to <code>NULL</code> as follows:</p>
<pre><code class="language-r">beta.dup &lt;- meffil.normalize.samples(
    norm.objects,
    just.beta=T, 
    cpglist.remove=qc.summary$bad.cpgs$name,
    dup.fun=NULL,
    verbose=T)
</code></pre>
<pre><code>## [read.idat] Mon Apr 17 00:42:32 2023 Reading data-epic2-demo/206891110001_R01C01_Grn.idat 
## [read.idat] Mon Apr 17 00:42:32 2023 Reading data-epic2-demo/206891110001_R01C01_Red.idat 
## [background.correct] Mon Apr 17 00:42:32 2023 background correction for dye = R 
## [background.correct] Mon Apr 17 00:42:34 2023 background correction for dye = G 
## [dye.bias.correct] Mon Apr 17 00:42:36 2023  
## [meffil.normalize.sample] Mon Apr 17 00:42:40 2023 Normalizing methylated and unmethylated signals.
</code></pre>
<p>We can then collapse the duplicate probes as follows:</p>
<pre><code class="language-r">beta.nodup &lt;- meffil.collapse.dups(beta.dup)
</code></pre>
<p>This function also has a <code>dup.fun</code> argument allowing the user
to specify a different summarizing function.</p>
<p>For example, cg06373096 appears
10 times.</p>
<pre><code class="language-r">quantile(beta.nodup[&quot;cg06373096&quot;,]-beta.meffil[&quot;cg06373096&quot;,],na.rm=T)
</code></pre>
<pre><code>##   0%  25%  50%  75% 100% 
##    0    0    0    0    0
</code></pre>
<pre><code class="language-r">quantile(colMedians(beta.dup[grepl(&quot;cg06373096&quot;,rownames(beta.dup)),],na.rm=T)-beta.meffil[&quot;cg06373096&quot;,],na.rm=T)
</code></pre>
<pre><code>##   0%  25%  50%  75% 100% 
##    0    0    0    0    0
</code></pre>
<p>Here are the correlations between these duplicates:</p>
<pre><code class="language-r">cor(t(beta.dup[grep(&quot;cg06373096&quot;,rownames(beta.dup)),]))
</code></pre>
<pre><code>##                  cg06373096 cg06373096_TC12 cg06373096_TC13 cg06373096_TC14
## cg06373096        1.0000000       0.8243699       0.7592742       0.7793531
## cg06373096_TC12   0.8243699       1.0000000       0.7897399       0.7397427
## cg06373096_TC13   0.7592742       0.7897399       1.0000000       0.7173957
## cg06373096_TC14   0.7793531       0.7397427       0.7173957       1.0000000
## cg06373096_TC15   0.8752993       0.7641724       0.8347072       0.8772968
## cg06373096_TC16   0.8825980       0.8154338       0.7713652       0.8267420
## cg06373096_TC17   0.7712993       0.7361161       0.6659681       0.6809560
## cg06373096_TC18   0.8289695       0.6955757       0.6189466       0.5692075
## cg06373096_TC19   0.7045560       0.7007004       0.8266329       0.7450534
## cg06373096_TC110  0.8118309       0.5930558       0.7121591       0.7932888
##                  cg06373096_TC15 cg06373096_TC16 cg06373096_TC17
## cg06373096             0.8752993       0.8825980       0.7712993
## cg06373096_TC12        0.7641724       0.8154338       0.7361161
## cg06373096_TC13        0.8347072       0.7713652       0.6659681
## cg06373096_TC14        0.8772968       0.8267420       0.6809560
## cg06373096_TC15        1.0000000       0.8176969       0.7212996
## cg06373096_TC16        0.8176969       1.0000000       0.6575714
## cg06373096_TC17        0.7212996       0.6575714       1.0000000
## cg06373096_TC18        0.6435677       0.8375359       0.5615424
## cg06373096_TC19        0.7625179       0.6663931       0.5967316
## cg06373096_TC110       0.8022731       0.7850622       0.6125044
##                  cg06373096_TC18 cg06373096_TC19 cg06373096_TC110
## cg06373096             0.8289695       0.7045560        0.8118309
## cg06373096_TC12        0.6955757       0.7007004        0.5930558
## cg06373096_TC13        0.6189466       0.8266329        0.7121591
## cg06373096_TC14        0.5692075       0.7450534        0.7932888
## cg06373096_TC15        0.6435677       0.7625179        0.8022731
## cg06373096_TC16        0.8375359       0.6663931        0.7850622
## cg06373096_TC17        0.5615424       0.5967316        0.6125044
## cg06373096_TC18        1.0000000       0.6419688        0.6092255
## cg06373096_TC19        0.6419688       1.0000000        0.6502503
## cg06373096_TC110       0.6092255       0.6502503        1.0000000
</code></pre>
<pre><code class="language-r">colMeans(cor(t(beta.dup[grep(&quot;cg06373096&quot;,rownames(beta.dup)),])))
</code></pre>
<pre><code>##       cg06373096  cg06373096_TC12  cg06373096_TC13  cg06373096_TC14 
##        0.8237550        0.7658907        0.7696189        0.7729036 
##  cg06373096_TC15  cg06373096_TC16  cg06373096_TC17  cg06373096_TC18 
##        0.8098831        0.8060399        0.7003989        0.7006540 
##  cg06373096_TC19 cg06373096_TC110 
##        0.7294804        0.7369650
</code></pre>
<p>How do matrices differ when we collapse
duplicates (<code>beta.meffil</code> and <code>beta.dup</code>)
or not (<code>beta.dup</code>)?</p>
<pre><code class="language-r">identical(rownames(beta.nodup), rownames(beta.meffil))
</code></pre>
<pre><code>## [1] TRUE
</code></pre>
<pre><code class="language-r">all(rownames(beta.nodup) %in% sub(&quot;_.*&quot;, &quot;&quot;, rownames(beta.dup)))
</code></pre>
<pre><code>## [1] TRUE
</code></pre>
<pre><code class="language-r">all(sub(&quot;_.*&quot;, &quot;&quot;, rownames(beta.dup)) %in% rownames(beta.nodup))
</code></pre>
<pre><code>## [1] TRUE
</code></pre>
<pre><code class="language-r">dim(beta.meffil)
</code></pre>
<pre><code>## [1] 917191     16
</code></pre>
<pre><code class="language-r">dim(beta.nodup)
</code></pre>
<pre><code>## [1] 917191     16
</code></pre>
<pre><code class="language-r">dim(beta.dup)
</code></pre>
<pre><code>## [1] 923547     16
</code></pre>


<script src="https://cdn.jsdelivr.net/combine/npm/@xiee/utils/js/center-img.min.js" async></script>
</body>

</html>
